design-document.md

```diff
@@ -14,4 +24,210 @@
 
 ## Open questions and system parameters
 
-## Cryptographic dependencies
+## Cryptographic components
+
+### The input-validation protocol
+
+**Notation.**
+Fix a finite field K. We let K^n denote the set of vectors over K of length n.
+We write the linear, s-way secret sharing of an element x of K^n as {x:1}, ...,
+{x:s}, where {x:i} is the share held by the i-th party. We write {x} as
+shorthand for the sequence {x:1}, ..., {x:s}.
+
+Prio combines standard linear secret sharing with a new type of
+probabilistically checkable proof (PCP) system, called a fully linear PCP. The
+input-input validation protocol can be described in terms of three main
+algorithms:
+
+1. pf := Prove(x) denotes generation of a proof pf of the validity of input x.
+   This algorithm is executed by the user.
+1. {vf:i} := Query({x:i}, {pf:i}, r) denotes computation of the verification
+   share {vf:i} for input share {x:i} and proof share {pf:i}. This algorithm is
+   executed by each of the aggregators; input r denotes the joint randomness
+   shared by all of the aggregators.
+1. b := Decide({vf}, r) denotes the execution of the decision procedure on input
+   shares {vf} and joint randomness r. The output b is a boolean indicating
+   whether the input is deemed valid. This algorithm is run by the leader.
```

Comment:
The threat model for Prio --- as it's described in the original paper and [BBC+19] --- considers **either** a malicious client (attacking soundness) **or** a malicious subset of aggregators (attacking privacy). In particular, soundness isn't guaranteed if any one of the aggregators is malicious; in theory it may be possible for a malicious client and aggregator to collude and break soundness.

There are techniques described in [BBC+19] that address the stronger threat model in which a malicious client may collude with a malicious aggregator. I think it's worth exploring how practical these techniques are. Meanwhile, thinking about attacks is a great idea.

Reply:
Maybe just flag this as an open issue?
```suggestion
   whether the input is deemed valid. This algorithm is run by the leader.
   
[[OPEN ISSUE: what can go wrong when the leader is responsible for verifying everything on behalf of all aggregators?]]
```

Reply:
Added this to "Thread model".